onnx/defs/tensor_proto_util.cc

```diff
@@ -69,6 +69,12 @@ namespace ONNX_NAMESPACE {
             " does not match the actual size",                                                                     \
             data.size());                                                                                          \
       }                                                                                                            \
+      if (tensor_proto->dims_size() != 0 && data.size() == 0) {                                                    \
```

Comment:
I don't think this is the right fix. This code just parses a TensorProto value. We can't return 0s as below. This code is called from many other places. The fix needed is in the caller. Specifically, we should not be calling `getInputData` in the implementation of [axisIsZero here](https://github.com/onnx/onnx/blob/96a0ca4374d2198944ff882bd273e64222b59cb9/onnx/defs/data_propagators.h#L36) ... we should be using the type of the input to figure out its rank, not look at its symbolic value.

Reply:
Created https://github.com/onnx/onnx/pull/6570

Reply:
I understand your perspective; however, in the axisIsZero function under the onnx/defs/data_propagators.h file, using type confirmation via rank can lead to errors in other units. I believe the issue lies with the parameter valueTypesByName in DataPropagationContextImpl, which prevents the use of getInputType to determine the rank.

     On the other hand, this error occurs in the PartialDataPropagationFunction located at line 583 of the onnx/defs/tensor/defs.cc file. This function depends on DataPropagationContext. According to the code design, the shape inference functionality that relies on DataPropagationContext requires using getInputData to determine the rank. The overall code logic appears to be structured in this way. I would appreciate any targeted insights or suggestions on addressing this issue. @gramalingam

Reply:
I understand your perspective; however, in the axisIsZero function under the onnx/defs/data_propagators.h file, using type confirmation via rank can lead to errors in other units. I believe the issue lies with the parameter valueTypesByName in DataPropagationContextImpl, which prevents the use of getInputType to determine the rank.

On the other hand, this error occurs in the PartialDataPropagationFunction located at line 583 of the onnx/defs/tensor/defs.cc file. This function depends on DataPropagationContext. According to the code design, the shape inference functionality that relies on DataPropagationContext requires using getInputData to determine the rank. The overall code logic appears to be structured in this way. I would appreciate any targeted insights or suggestions on addressing this issue. @gramalingam

Reply:
getInputData in DataPropagationContext carries only information about the values of the elements of an input tensor, but no information about the shape of the input tensor. So, it is not possible to get information about the shape/rank of an input tensor from getInputData (in general). It encodes the value information using a TensorShapeProto so that, eg., a value like `(1, 4, A, B)` can be used to indicate that the tensor contains the known values 1 and 4, followed by two unknown values denotes A and B. But we don't know if this is a 1D tensor or 2D tensor. I suspect this bug may have been caused by confusing this TensorShapeProto (which describes the values of the input tensor) with the shape of the input tensor.

Reply:
TensorShapeProto describes the dimension information of a tensor. For example, a tensor with dimensions (B, 1, 255, 255) represents a 4-dimensional tensor where B, 1, 255, and 255 specify the size of each dimension. Therefore, I believe that defining the return value of the getInputData function in the DataPropagationContextImpl structure as TensorShapeProto is appropriate. From a code design perspective, the sole purpose of the getInputData function in DataPropagationContextImpl is to return the dimensional information of the input tensor. This differs fundamentally from the design of the getInputData function in InferenceContextImpl, which serves a different primary function.

Reply:
**For example, the following code snippet demonstrates that TensorShapeProto represents the dimensional information of a tensor.**
```python
import onnx
from onnx.helper import (
    make_model, make_node, make_tensor, make_graph,
    make_tensor_value_info)
from onnx import helper, numpy_helper, TensorProto, TensorShapeProto, ValueInfoProto
import numpy as np

def create_and_save_model():
    input_a_shape = TensorShapeProto(
        dim=[
            TensorShapeProto.Dimension(dim_param="batch"),  
            TensorShapeProto.Dimension(dim_value=3),        
        ]
    )
    
    input_b_shape = TensorShapeProto(
        dim=[
            TensorShapeProto.Dimension(dim_param="batch"), 
            TensorShapeProto.Dimension(dim_value=3),        
        ]
    )

    input_a_type = onnx.TypeProto()
    input_a_type.tensor_type.elem_type = TensorProto.FLOAT
    input_a_type.tensor_type.shape.CopyFrom(input_a_shape)

    input_b_type = onnx.TypeProto()
    input_b_type.tensor_type.elem_type = TensorProto.FLOAT
    input_b_type.tensor_type.shape.CopyFrom(input_b_shape)

    input_a = ValueInfoProto(
        name="input_a",
        type=input_a_type
    )

    input_b = ValueInfoProto(
        name="input_b",
        type=input_b_type
    )

    output = ValueInfoProto(
        name="output",
        type=input_a_type
    )

    node_def = helper.make_node(
        'Add',                     
        ['input_a', 'input_b'],    
        ['output'],               
        name='add_node'           
    )

    constant_tensor = numpy_helper.from_array(
        np.array([[1.0, 2.0, 3.0]], dtype=np.float32),
        name="constant"
    )

    graph_def = helper.make_graph(
        nodes=[node_def],          
        name='add_graph',         
        inputs=[input_a, input_b],
        outputs=[output],          
        initializer=[constant_tensor] 
    )

    model_def = helper.make_model(
        graph_def,
        producer_name='onnx_demo',
        opset_imports=[helper.make_opsetid("", 13)]
    )

    onnx.checker.check_model(model_def)

    onnx.save(model_def, "add_model.onnx")

    return model_def

model = create_and_save_model()
```

![image](https://github.com/user-attachments/assets/cfdb6456-2bea-4d20-b615-fe5140b9fddf)